webapp/app/forms/fields.py

```diff
@@ -70,18 +70,32 @@ class IdNrWidget(MultipleInputFieldWidget):
 
 
 class IdNrField(SteuerlotseStringField):
+    """
+        Field to store the IdNr in four separate input fields.
+
+        We get the formdata as a list of four strings (e.g. ['04', '452', '397', '687'])
+        but want to handle it in the rest of the program as one string (e.g. '04452397687').
+        At the same time, in case of a validation error (e.g. for ['04', '452', '3', '687']) we want to keep the order
+        of inputs to not confuse the user. Thus, we only concatenate the strings to one string on succeeded validation
+        in post_validate().
+        Once the input validates, we can be sure to have the complete, valid string in our data and
+        can split it into the expected chunks as seen in _value().
+    """
     def __init__(self, label='', validators=None, **kwargs):
         super(IdNrField, self).__init__(label, validators, **kwargs)
         self.widget = IdNrWidget()
 
     def process_formdata(self, valuelist):
+        # The formdata (from the request) is written to self.data as is (as a list of inputted strings)
         if valuelist:
             self.data = valuelist
         elif self.data is None:
             self.data = []
 
     def _value(self):
-        # In case the validation was not successful, we already have the data as a list of values.
+        """ Returns the representation of data as needed by the widget. In this case: a list of strings. """
```

Comment:
This is always only called after a successful validation, correct?

Reply:
No, it is not. This is the method that is called whenever the form wants to know what to display in the widget. So whenever the field is rendered, this is called.

Reply:
Okidoki